Source/WebGPU/WGSL/AST/ASTBuilder.cpp

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ASTBuilder.h"
+
+namespace WGSL::AST {
+
+DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(WGSLAST);
+
+Builder::Builder(Builder&& other)
+{
+    m_arena = std::exchange(other.m_arena, nullptr);
+    m_arenaEnd = std::exchange(other.m_arenaEnd, nullptr);
+    m_arenas = WTFMove(other.m_arenas);
+    m_nodes = WTFMove(other.m_nodes);
+}
+
+inline uint8_t* Builder::arena()
+{
+    ASSERT(m_arenaEnd);
+    return m_arenaEnd - arenaSize;
+}
+
+Builder::~Builder()
+{
+    size_t size = m_nodes.size();
+    for (size_t i = 0; i < size; ++i)
+        m_nodes[i]->~Node();
+
+    if (m_arenaEnd)
+        WGSLASTMalloc::free(arena());
+
+    size = m_arenas.size();
+    for (size_t i = 0; i < size; ++i)
+        WGSLASTMalloc::free(m_arenas[i]);
+}
+
+void Builder::allocateArena()
+{
+    if (m_arenaEnd)
+        m_arenas.append(arena());
+
+    m_arena = static_cast<uint8_t*>(WGSLASTMalloc::malloc(arenaSize));
```

Comment:
Surely we shouldn't be using manual malloc here?

Reply:
Why? That's how it works in JSC's ParserArena

Reply:
it's still an anti-pattern :/

What would it take to use `unique_ptr` instead?

Reply:
I don't think it's an anti pattern to call malloc from an allocator... I guess we could use an unique_ptr to an uint8_t array, but we'd lose the heap identifier. I can see why you'd prefer that in other scenarios, but in this case I don't see the benefit. We have one allocation and one deallocation site, with a very well defined lifecycle, which will be allocating large amounts of memory.

Reply:
I switch to `MallocPtr`, hopefully that's a reasonable middle ground?